File: shellsort.py
def shellsort(v, n):
    data = [v[:]]
    gap = int(n/2)
    while gap > 0:
        for i in range(gap, n):
            j = i - gap
            while (j >= 0) and (v[j] > v[j+gap]):
                temp = v[j]
                v[j] = v[j+gap]
                v[j+gap] = temp
                j -= gap
                data.append(v[:])
        gap = int(gap / 2)
    return data


if __name__ == '__main__':
    import matplotlib.pyplot as plt

    v = [2, 6, 7, 5, 7, 9, 0, 10, 10, 1]
    n = len(v)

    d = shellsort(v, n)

    fig, ax = plt.subplots(nrows=3, ncols=3)
    ax = ax.flatten()
    for i, j in enumerate(d):
        ax[i].bar(list(range(0, len(j))), j)
    plt.show()

File: shellsort.c
#include <stdio.h>

void shellsort(int v[], int n);

int main(void) {
    int i, n;
    int v[] = {2, 6, 7, 5, 7 , 9, 0 , 10, 10, 1};

    n = 10;
    shellsort(v, n);
    
    for (i = 0; i < n; ++i)
        printf("%d, ", v[i]); 
}

/* sort in increasing order */
void shellsort(int v[], int n) {
    int gap, i, j, temp;

    for (gap = n/2; gap > 0; gap/=2)
        for (i = gap; i < n; i++)
            for (j=i-gap; j>=0 && v[j]>v[j+gap]; j-=gap) {
                temp = v[j];
                v[j] = v[j+gap];
                v[j+gap] = temp;
            }
}
